service/storage_proxy.cc

```diff
@@ -723,12 +723,48 @@ class storage_proxy::remote {
         auto cmd = make_lw_shared<query::read_command>(std::move(cmd1));
         auto src_ip = src_addr.addr;
         auto timeout = t ? *t : db::no_timeout;
+
+        // Pull a schema from the coordinator. There are two cases:
+        // 1. a mixed nodes cluster, cmd->schema_version is a table schema, a table schema is pulled
+        // 2. all new nodes cluster, cmd->schema_version is a reversed table schema, a reversed table schema is pulled
         auto f_s = co_await coroutine::as_future(get_schema_for_read(cmd->schema_version, std::move(src_addr), timeout));
         if (f_s.failed()) {
             co_return co_await encode_replica_exception_for_rpc<Result>(p->features(), f_s.get_exception());
         }
         schema_ptr s = f_s.get();
 
+        // Verify whether read_command is provided in legacy or native reversed format by comparing
+        // schema version with table schema version. There are three options:
+        // 1. Versions are the same -> legacy format
+        // 2. schema version is equal to reversed table schema -> native format
+        // 3. table schema is outdated
+        static auto is_read_command_in_reversed_legacy_format = [] (const query::read_command& cmd, schema_ptr table_schema) -> std::optional<bool> {
+            if (cmd.schema_version == table_schema->version()) {
+                return true;
+            }
+
+            if (cmd.schema_version == reversed(table_schema->version())) {
+                return false;
+            }
+
+            return std::nullopt;
+        };
+
+        // Detect whether a transformation from legacy reverse format into native reverse
+        // format is necessary before executing the read_command. That happens when the
+        // native_reverse_queries feature is turned off. A custer has mixed nodes.
+        bool format_reverse_required = false;
+
+        // Check if we have a reversed query
+        if (cmd->slice.is_reversed()) {
+            // Compare the schema version provided by read_command with the table schema replica holds.
+            // Note that with the get_schema_for_read call above with reversed table schema, it pulls
+            // the table schema as well. Thus table_schema is always up-to-date.
+            auto table_schema = p->local_db().find_schema(cmd->cf_id);
+            format_reverse_required = *is_read_command_in_reversed_legacy_format(*cmd, table_schema);
```

Comment:
`is_read_command_in_reversed_legacy_format()` returns an optional and has a return-path which returns a disegaged-one, yet you undconditionally dereference it here.
Either make it not return an optional, or check the return value before dereferencing. The current code is prone to stepping on SEFAULT (`std::optional::operator*()` doesn't check that the optional is engaged before defererencing).